File: DSA/CPP/Divide And Conquer/quickSort.cpp
#include <bits/stdc++.h>
using namespace std;


int partition(int arr[], int start, int end){

    int pivotIndex = start;
    int pivotElem = arr[start];

    // left side of pivot  >= privotELement
    // right side of pivot < privotElement

    while(start <= end){

        if (arr[start] <= pivotElem){
            start++;
        }

        else if (arr[end] > pivotElem){
            end--;
        }

        else{
            swap(arr[start], arr[end]);
        }
    }

    swap(arr[pivotIndex], arr[end]);
    return end;

}

// 2 5 3 1 6 8 9 7

void quickSort(int arr[], int start, int end){

    // signle is always sorted
    if (start >= end) 
        return; 

    int index = partition(arr, start, end);

    //  index = 5 - first element in its correct position
    // what about elements from 0 - index - 1 & index  + 1, to n - 1
    quickSort(arr, start, index - 1);
    quickSort(arr, index + 1, end);

}


int main(){

    int n = 8;
    int arr[] = {7, 4, 2, 1, 0, 9, 3, 6};

    for (int i = 0; i < n; i++){
        cout << arr[i] << " ";
    }
    cout << "\n";

    quickSort(arr, 0, n - 1);
    
    for (int i = 0; i < n; i++){
        cout << arr[i] << " ";
    }
    cout << "\n";



    return 0;
}

File: DSA/JAVA/Divide And Conquer/QuickSort.java
import java.util.Arrays;

public class QuickSortDemo {

    // Partition function
    public static int partition(int[] arr, int start, int end) {
        int pivotIndex = start;
        int pivotElement = arr[start];

        while (start <= end) {
            if (arr[start] <= pivotElement) {
                start++;
            } else if (arr[end] > pivotElement) {
                end--;
            } else {
                // Swap elements at start and end
                int temp = arr[start];
                arr[start] = arr[end];
                arr[end] = temp;
            }
        }

        // Swap pivot with the correct position
        int temp = arr[pivotIndex];
        arr[pivotIndex] = arr[end];
        arr[end] = temp;

        return end; // Return the final position of the pivot
    }

    // QuickSort function
    public static void quickSort(int[] arr, int start, int end) {
        if (start >= end) {
            return;
        }

        int pivotPosition = partition(arr, start, end);
        quickSort(arr, start, pivotPosition - 1);
        quickSort(arr, pivotPosition + 1, end);
    }

    // Main function
    public static void main(String[] args) {
        int[] arr = {7, 4, 2, 1, 0, 9, 3, 6};

        System.out.println("Before Sorting:");
        System.out.println(Arrays.toString(arr));

        quickSort(arr, 0, arr.length - 1);

        System.out.println("After Sorting:");
        System.out.println(Arrays.toString(arr));
    }
}